### DISCUSSION TITLE
Differences between fastapi.query and pydantic.conint for numeric validation

### DISCUSSION THREAD

**Discussion by LucasLev** [fastapi/fastapi#9607] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Query
from pydantic import conint

app = FastAPI()


@app.get("/positive_query")
def want_positive(positive: int = Query(default=..., gt=0)):
    return f"{positive} is positive !"


@app.get("/positive_query_pydantic")
def want_positive_2(positive: conint(gt=0)):
    return f"{positive} is positive ! (but with pydantic !)"
```

### Description

What is the most natural way to ensure an int from a query respects certains condition ?

I've been using ``pydantic.conint`` as it seems the most natural way to coerce an int (and the same way we coerce it in the models), but some others members of my project used the ``fastapi.query`` form described in the doc. 

What are the added benefits of using ``fastapi.Query`` in this case ? As far as I can tell, there is no difference in the output and query is more lengthy in the code for a simple task.

(nb: I saw the new annotated syntax, but it doesn't change much about the problem)

### Operating System

Linux

### FastAPI Version

0.76.0

### Python Version

3.10.0

**Comment by esrefzeki** [comment]
Using Pydantic's data types (such as 'conint') along with FastAPI's 'Query' function provides a more declarative and reusable approach. By defining a Pydantic model with validation rules, you can leverage the power of Pydantic's validation and coercion capabilities. This allows you to define complex validation rules, apply defaults, and benefit from Pydantic's rich ecosystem of validation decorators, such as 'gt' (greater than) in your example.

On the other hand, using FastAPI's 'Query' directly provides a more concise and straightforward approach. It allows you to define validation rules directly on the query parameter itself, without the need for a separate Pydantic model. This approach might be more suitable for simple validation cases where you don't need the additional capabilities provided by Pydantic.

However, if you have more complex validation needs or want to reuse the validation rules across multiple endpoints, using Pydantic's data types is a better choice. It gives you the flexibility to define rich validation rules, apply defaults, and take advantage of Pydantic's validation decorators.

**Comment by ahmedabdou14** [comment, marked as answer]
Actually FastApi internally uses pydantic for data validation. the `Query(..., gt=0)`  uses the same validation used by `contint`. It might be opinionated, but I personally prefer using fastapi `Query`  and specifying the type of query `n: int = Query(..., gt=0)` since it combines the best of both worlds
- Pydantic validation
- Clear and concise syntax _(especially when method signature contains a lot of params)_
- FastApi Query customization like  `Query(..., gt=0, deprecated=True, description="", alias="id")`

**Reply by LucasLev** [reply]
Very clear, thank you !

**Reply by nerdstrike** [reply]
Annoyingly you can still import `PositiveInt` from `pydantic`, but it's from the v1 API and silently fails to do the job when specified as part of a route Query.
